In [172]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stat
import pylab 

In [330]:
df = pd.read_csv('Bengaluru_House_Data.csv')

In [331]:
print(f" Shape have {df.shape[0]} rows and {df.shape[1]} columns")

 Shape have 13320 rows and 9 columns


In [509]:
df.sample(5)

location  size  total_sqft  bath  balcony  price  Whitefield  \
6716     Lingarajapuram     1       450.0   1.0      2.0   60.0           0   
7003  Nrupathunga Nagar     7      1200.0   5.0      0.0   80.0           0   
2459         Rayasandra     7       700.0   6.0      2.0   85.0           0   
4619         Whitefield     3      1700.0   3.0      2.0   95.0           1   
6989    Electronic City     3      1631.0   3.0      2.0   80.0           0   

      Sarjapur  Road  Electronic City  Kanakpura Road  ...  Hebbal Kempapura  \
6716               0                0               0  ...                 0   
7003               0                0               0  ...                 0   
2459               0                0               0  ...                 0   
4619               0                0               0  ...                 0   
6989               0                1               0  ...                 0   

      Basavangudi  Kumaraswami Layout  Mahadevpura  Anekal  Ambedkar Nagar  \
6716            0                   0            0       0               0   
7003            0                   0            0       0               0   
2459            0                   0            0       0               0   
4619            0                   0            0       0               0   
6989            0                   0            0       0               0   

      Doddathoguru  Old Airport Road  Chikkalasandra  Kammasandra  
6716             0                 0               0            0  
7003             0                 0               0            0  
2459             0                 0               0            0  
4619             0                 0               0            0  
6989             0                 0               0            0  

[5 rows x 106 columns]

In [333]:
df.isnull().mean()*100

area_type        0.000000
availability     0.000000
location         0.007508
size             0.120120
society         41.306306
total_sqft       0.000000
bath             0.548048
balcony          4.572072
price            0.000000
dtype: float64

In [334]:
df.duplicated().sum()

529

In [335]:
df.drop_duplicates(inplace=True)

In [336]:
df.shape

(12791, 9)

In [337]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12791 entries, 0 to 13318
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   area_type     12791 non-null  object 
 1   availability  12791 non-null  object 
 2   location      12790 non-null  object 
 3   size          12775 non-null  object 
 4   society       7463 non-null   object 
 5   total_sqft    12791 non-null  object 
 6   bath          12718 non-null  float64
 7   balcony       12186 non-null  float64
 8   price         12791 non-null  float64
dtypes: float64(3), object(6)
memory usage: 999.3+ KB


In [338]:
df.drop(columns=['society','area_type','availability'],inplace=True)

In [339]:
df.head()

location       size total_sqft  bath  balcony   price
0  Electronic City Phase II      2 BHK       1056   2.0      1.0   39.07
1          Chikka Tirupathi  4 Bedroom       2600   5.0      3.0  120.00
2               Uttarahalli      3 BHK       1440   2.0      3.0   62.00
3        Lingadheeranahalli      3 BHK       1521   3.0      1.0   95.00
4                  Kothanur      2 BHK       1200   2.0      1.0   51.00

# report 
* society having more null value we drop the society feature 
* data have duplicate 529 value we drop value 
* we drop 'society','area_type','availability' feature because this not much usefull feature

In [340]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12791 entries, 0 to 13318
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   location    12790 non-null  object 
 1   size        12775 non-null  object 
 2   total_sqft  12791 non-null  object 
 3   bath        12718 non-null  float64
 4   balcony     12186 non-null  float64
 5   price       12791 non-null  float64
dtypes: float64(3), object(3)
memory usage: 699.5+ KB


In [341]:
print('Unique value count in each feature')
print()
data_columns = ['location','size','total_sqft','bath','balcony','price']
for col in data_columns:
    u = len(df[col].unique())
    print(col,"------>",u)

Unique value count in each feature

location ------> 1306
size ------> 32
total_sqft ------> 2117
bath ------> 20
balcony ------> 5
price ------> 1994


In [342]:
df.isnull().sum()

location        1
size           16
total_sqft      0
bath           73
balcony       605
price           0
dtype: int64

In [343]:
df['bath']= df['bath'].fillna(df['bath'].median())
df['balcony']= df['balcony'].fillna(df['balcony'].median())
df['location']= df['location'].fillna(df['location'].mode()[0])
df['size']= df['size'].fillna(df['size'].mode()[0])

In [344]:
df.isnull().sum()

location      0
size          0
total_sqft    0
bath          0
balcony       0
price         0
dtype: int64

In [345]:
df['size'] = df['size'].str.replace('BHK','').str.replace('Bedroom','').str.replace('RK','').str.replace(" ","")

In [346]:
df['size'].info()

<class 'pandas.core.series.Series'>
Int64Index: 12791 entries, 0 to 13318
Series name: size
Non-Null Count  Dtype 
--------------  ----- 
12791 non-null  object
dtypes: object(1)
memory usage: 199.9+ KB


In [347]:
df.head()

location size total_sqft  bath  balcony   price
0  Electronic City Phase II    2       1056   2.0      1.0   39.07
1          Chikka Tirupathi    4       2600   5.0      3.0  120.00
2               Uttarahalli    3       1440   2.0      3.0   62.00
3        Lingadheeranahalli    3       1521   3.0      1.0   95.00
4                  Kothanur    2       1200   2.0      1.0   51.00

In [348]:
df['size'] = df['size'].astype('int')

In [349]:
invalid_rows = df[df['total_sqft'].str.contains('Sq. Meter')]
df.drop(invalid_rows.index, inplace=True) 
invalid_rows = df[df['total_sqft'].str.contains('Perch')]
df.drop(invalid_rows.index, inplace=True) 
invalid_rows = df[df['total_sqft'].str.contains('Sq. Yards')]
df.drop(invalid_rows.index, inplace=True) 
invalid_rows = df[df['total_sqft'].str.contains('Acres')]
df.drop(invalid_rows.index, inplace=True) 
invalid_rows = df[df['total_sqft'].str.contains('Grounds')]
df.drop(invalid_rows.index, inplace=True) 
invalid_rows = df[df['total_sqft'].str.contains('Cents')]
df.drop(invalid_rows.index, inplace=True) 
invalid_rows = df[df['total_sqft'].str.contains('Guntha')]
df.drop(invalid_rows.index, inplace=True) 

In [350]:
def convert_sqft(values):
        if "-" in values:
            start , end = values.split("-")
            return (float(start) + float(end))/2
        else:
            return float(values)
df['total_sqft'] = df['total_sqft'].apply(convert_sqft)

In [351]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12745 entries, 0 to 13318
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   location    12745 non-null  object 
 1   size        12745 non-null  int32  
 2   total_sqft  12745 non-null  float64
 3   bath        12745 non-null  float64
 4   balcony     12745 non-null  float64
 5   price       12745 non-null  float64
dtypes: float64(4), int32(1), object(1)
memory usage: 647.2+ KB


In [352]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
from sklearn.preprocessing import StandardScaler,MinMaxScaler,OneHotEncoder

In [486]:
top_100 =[x for x in df['location'].value_counts().sort_values(ascending=False).head(100).index]

In [492]:
for label in top_100:
    df[label] = np.where(df['location']==label,1,0)
df.head()

location  size  total_sqft  bath  balcony   price  \
0  Electronic City Phase II     2      1056.0   2.0      1.0   39.07   
1          Chikka Tirupathi     4      2600.0   5.0      3.0  120.00   
2               Uttarahalli     3      1440.0   2.0      3.0   62.00   
3        Lingadheeranahalli     3      1521.0   3.0      1.0   95.00   
4                  Kothanur     2      1200.0   2.0      1.0   51.00   

   Whitefield  Sarjapur  Road  Electronic City  Kanakpura Road  ...  \
0           0               0                0               0  ...   
1           0               0                0               0  ...   
2           0               0                0               0  ...   
3           0               0                0               0  ...   
4           0               0                0               0  ...   

   Hebbal Kempapura  Basavangudi  Kumaraswami Layout  Mahadevpura  Anekal  \
0                 0            0                   0            0       0   
1                 0            0                   0            0       0   
2                 0            0                   0            0       0   
3                 0            0                   0            0       0   
4                 0            0                   0            0       0   

   Ambedkar Nagar  Doddathoguru  Old Airport Road  Chikkalasandra  Kammasandra  
0               0             0                 0               0            0  
1               0             0                 0               0            0  
2               0             0                 0               0            0  
3               0             0                 0               0            0  
4               0             0                 0               0            0  

[5 rows x 106 columns]

In [493]:
X = df.drop(columns=['location','price'])
y = df['price']

In [494]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=34)

In [502]:
scaling = StandardScaler()
X_train_scaled = scaling.fit_transform(X_train,y_train)
X_test_scaled = scaling.transform(X_test)

In [506]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.svm import SVR
lr =  SVR()
lr.fit(X_train,y_train)

SVR()

In [507]:
y_pred = lr.predict(X_test)

In [508]:
r2_score(y_test,y_pred)

0.3235970445064321